File: vanhak/SortColumnsOfCsvFile/Challenge.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SortColumnsOfCsvFile
{
    public class Challenge
    {
        private const char COLUMN_JOIN = ',';
        private const char LINE_SEPARATOR = '\n';
        public static string SortCsvColumns(string csvData)
        {
            if (string.IsNullOrWhiteSpace(csvData))
            {
                return null;
            }
            var columns = ToColumns(csvData);
            var gfdg = ToCsv(columns.OrderBy(m => m.Values.First()).ToList());
            return gfdg;
        }

        private static List<Column> ToColumns(string csvData)
        {
            var rows = csvData.Split(LINE_SEPARATOR).Select(s => s.Split(COLUMN_JOIN)).ToArray();
            var columnsCount = rows.First().Count();
            var rowsCount = rows.Count();
            var columns = new List<Column>();
            for (var i = 0; i < columnsCount; i++)
            {
                var c = new Column
                {
                    Values = new string[rowsCount]
                };
                for (var j = 0; j < rowsCount; j++)
                {
                    var v = rows[j][i];
                    c.Values[j] = string.IsNullOrWhiteSpace(v) ? string.Empty : v;
                }
                columns.Add(c);
            }
            return columns;
        }

        private static string ToCsv(List<Column> columns)
        {
            var r = new List<string>();
            var columnsCount = columns.Count();
            var rowsCount = columns.First().Values.Count();
            for (var i = 0; i < rowsCount; i++)
            {
                var s = new List<string>();
                for (var j = 0; j < columnsCount; j++)
                {
                    s.Add(columns[j].Values[i]);
                }
                r.Add(string.Join(COLUMN_JOIN, s));
            }
            return string.Join(LINE_SEPARATOR, r);
        }

        public class Column
        {
            public string[] Values { get; set; }
        }
    }
}

File: vanhak/sortColumnsOfCsvFile_Golang/challenge_test/challenge.go
package challenge

import (
	"sort"
	"strings"
)

const COLUMN_JOIN = ","
const LINE_SEPARATOR = "\n"

type Column struct {
	Values []string
}

func SortCsvColumns(csvData string) string {
	if csvData == "" {
		return ""
	}
	columns := toColumns(csvData)
	sort.Slice(columns, func(i, j int) bool {
		return strings.ToLower(columns[i].Values[0]) < strings.ToLower(columns[j].Values[0])
	})
	return toCsv(columns)
}

func toColumns(csvData string) []Column {
	lines := strings.Split(csvData, LINE_SEPARATOR)
	rows := make([][]string, len(lines))
	for i, l := range lines {
		rows[i] = strings.Split(l, COLUMN_JOIN)
	}
	columnCount := len(rows[0])
	rowsCount := len(rows)
	var columns = []Column{}
	for i := 0; i < columnCount; i++ {
		c := Column{
			Values: make([]string, rowsCount),
		}
		for j := 0; j < rowsCount; j++ {
			var v = rows[j][i]
			if v == " " || v == "" {
				c.Values[j] = ""
			} else {
				c.Values[j] = v
			}
		}
		columns = append(columns, c)
	}
	return columns
}

func toCsv(columns []Column) string {
	r := []string{}
	columnCount := len(columns)
	rowsCount := len(columns[0].Values)
	for i := 0; i < rowsCount; i++ {
		s := []string{}
		for j := 0; j < columnCount; j++ {
			s = append(s, columns[j].Values[i])
		}
		r = append(r, strings.Join(s, COLUMN_JOIN))
	}
	return strings.Join(r, LINE_SEPARATOR)
}
